### ISSUE TITLE
Brainstorm parsing optional arguments

### ISSUE THREAD

**Issue by maringuu** [ifreund/river#102]
For #101 to be implemented we could think about a nice way to parse optional arguments. After some brainstorming I though of maybe using an `enum` and an option_mask which specifies options in the same way that tagmask specifies tags. This enum would be generated in the file which implements the command.
This enum could be generated from a nice descriptive list of options and their default value.
This could look something like this: 
`map.zig`
```zig
const OptionalArgs: enum(u32) = //Generated somehow.

const arglist = [_]struct {
    name: []const u8,
    default_value: bool, // Optionals can only be true or false
    name_in_enum: []u8, // The names of the entries in the enum have to come somewhere 
                    // Perhabs we could generate them from name
                    //but this would make it tricky to have commands that contain illegal characters for zig enum members
}{... // The args specified
}
```

Edit: The default value will be needed in the parsing functions which needs to determine what should be set when the option is not passed.
What are your thoughts and ideas about this?

**Comment by ifreund** [comment]
Zig's metaprogramming is very powerful. Something like this should do quite nicely:

```zig
fn parse(args: []const []const u8, T: type) T {
    var ret = T{};

    for (args) |arg| {
        inline for (@typeInfo(T).Struct.fields) |field| {
            if (std.mem.eql(u8, arg, field.name)) {
                switch (@typeInfo(field.field_type)) {
                    .Int => |int| {
                        @field(ret, field.name) = try std.fmt.parseInt(int, arg, 10);
                    },
                    // other supported types
                    else => @compileError("unsupported option type: " ++ @typeName(field.field_type)),
                }
            }
        }
    }
}

// Usage

fn foo() void {
    const options = parse(args, struct {
        @"-no-funny-stuff": bool = true,
        @"-blah-val": u32 = 42,
    });
}
```

That of course is not working code, but it demonstrates the idea well enough.

**Comment by maringuu** [comment]
That looks indeed very powerful. That looks like a good idea.
That optional arg can have other types than `bool` is indeed true and rules out the `enum` idea.

Would you like to implement this? This is your idea after all. If not I'd like to do it to get some expericene with zig meta programming.
When implemented I'll give #12 a try to have the new generic version be used in more than one place.

**Comment by ifreund** [comment]
> Would you like to implement this? This is your idea after all. If not I'd like to do it to get some expericene with zig meta programming.

Go for it, I have too many things I want to do and such little time :)
